### PR TITLE
[SandboxIR] Implement PHINodes

### PR DESCRIPTION
**Pull request by Sterling-Augustine**
This patch implements sandboxir::PHINode which mirrors llvm::PHINode.

Based almost entirely on work by vporpo.

### CODE REVIEW

File: llvm/include/llvm/SandboxIR/SandboxIR.h [hunk 1]
```diff
@@ -1290,6 +1292,97 @@ class GetElementPtrInst final : public Instruction {
 #endif
 };
 
+class PHINode final : public Instruction {
+  /// Use Context::createPHINode(). Don't call the
+  /// constructor directly.
+  PHINode(llvm::PHINode *PHI, Context &Ctx)
+      : Instruction(ClassID::PHI, Opcode::PHI, PHI, Ctx) {}
+  friend Context; // for SBPHINode()
```
Comment (vporpo) [inline llvm/include/llvm/SandboxIR/SandboxIR.h:1300]: SBPHINode -> PHINode

File: llvm/lib/SandboxIR/SandboxIR.cpp [hunk 2]
```diff
@@ -1301,6 +1407,11 @@ Context::createGetElementPtrInst(llvm::GetElementPtrInst *I) {
   return cast<GetElementPtrInst>(registerValue(std::move(NewPtr)));
 }
 
+PHINode *Context::createPHINode(llvm::PHINode *I) {
+  auto NewPtr = std::unique_ptr<PHINode>(new PHINode(I, *this));
+  return cast<PHINode>(registerValue(std::move(NewPtr)));
```
Comment (vporpo) [inline llvm/lib/SandboxIR/SandboxIR.cpp:1412]: nit: Perehaps `return cast<PHINode>(registerValue(std::unique_ptr<PHINode>(new PHINode(I, *this))));` ?

File: llvm/include/llvm/SandboxIR/SandboxIR.h [hunk 3]
```diff
@@ -1290,6 +1292,97 @@ class GetElementPtrInst final : public Instruction {
...
+  int getBasicBlockIndex(const BasicBlock *BB) const;
+  Value *getIncomingValueForBlock(const BasicBlock *BB) const;
+
+  Value *hasConstantValue() const;
+
+  bool hasConstantOrUndefValue() const {
+    return cast<llvm::PHINode>(Val)->hasConstantOrUndefValue();
+  }
+  bool isComplete() const { return cast<llvm::PHINode>(Val)->isComplete(); }
+
```
Comment (vporpo) [inline llvm/include/llvm/SandboxIR/SandboxIR.h:1376]: Some member functions have not been implemented yet, like `replaceIncomingBlockWith()`, `copyIncomingBlocks()`, `removeIncomingValueIf()`. Could you add a TODO in the code ?

**Review by vporpo — APPROVED** [review]
Just some minor comments, looks good.

File: llvm/include/llvm/SandboxIR/SandboxIR.h [hunk 4]
```diff
@@ -1290,6 +1292,97 @@ class GetElementPtrInst final : public Instruction {
 #endif
 };
 
+class PHINode final : public Instruction {
+  /// Use Context::createPHINode(). Don't call the
```
[the remark this answers is not in the document]
Comment (Sterling-Augustine) [reply llvm/include/llvm/SandboxIR/SandboxIR.h:1296]: done

File: llvm/include/llvm/SandboxIR/SandboxIR.h (hunk 1 above)
Comment (Sterling-Augustine) [reply llvm/include/llvm/SandboxIR/SandboxIR.h:1300]: done

File: llvm/lib/SandboxIR/SandboxIR.cpp (hunk 2 above)
Comment (Sterling-Augustine) [reply llvm/lib/SandboxIR/SandboxIR.cpp:1412]: That ends up being different from the others, like on the code just above this one. Should we change them all?

File: llvm/include/llvm/SandboxIR/SandboxIR.h (hunk 3 above)
Comment (Sterling-Augustine) [reply llvm/include/llvm/SandboxIR/SandboxIR.h:1376]: done

File: llvm/lib/SandboxIR/SandboxIR.cpp (hunk 2 above)
Comment (vporpo) [reply llvm/lib/SandboxIR/SandboxIR.cpp:1412]: Sure, your call. Either is fine with me.

Comment (vporpo) [reply llvm/lib/SandboxIR/SandboxIR.cpp:1412]: I mean we can keep it as it is too, I don't feel strongly about changing them.

**Pull request merged**